packages/blocks/src/lib/navigation-map/components/obstacle.svelte

```diff
@@ -1,26 +1,145 @@
 <script lang="ts">
 import { theme } from '@viamrobotics/prime-core/theme';
 import * as THREE from 'three';
-import { T } from '@threlte/core';
-import type { Obstacle } from '$lib';
-import { view, hovered, obstacleNavItems } from '../stores';
+import { T, createRawEventDispatcher } from '@threlte/core';
+import type { MapMouseEvent } from 'maplibre-gl';
+import { useMapLibre, type Obstacle, useMapLibreEvent } from '$lib';
+import { view, hovered, selected, environment } from '../stores';
 import AxesHelper from './axes-helper.svelte';
 
 /** An obstacle to render. */
 export let obstacle: Obstacle;
 
+interface $$Events extends Record<string, unknown> {
+  /** Fired when obstacles are created, destroyed, or edited. */
+  update: Obstacle;
+}
+
+const dispatch = createRawEventDispatcher<$$Events>();
+const { map } = useMapLibre();
+
 let material: THREE.MeshPhongMaterial;
 
+let pointerdownTheta = 0;
+let pointerdownRadius = 0;
+let pointerdownLength = 0;
+let pointerdownWidth = 0;
+let pointerdownHeight = 0;
+
+let draggingObstacle = false;
+
+const pointermove = new THREE.Vector2();
+const pointerdown = new THREE.Vector2();
+
 const handleGeometryCreate = ({ ref }: { ref: THREE.BufferGeometry }) => {
   ref.rotateX(-Math.PI / 2);
 };
 
+const handlePointerDown = (name: string) => {
+  $selected = name;
+
+  if (debugMode) {
+    return;
+  }
+
+  map.dragPan.disable();
+  draggingObstacle = true;
+};
+
+const handlePointerMove = (event: MapMouseEvent) => {
+  if ($selected === null) {
+    return;
+  }
+
+  // Rotate
+  if (event.originalEvent.metaKey) {
+    pointermove.set(event.point.x, event.point.y);
+    pointermove.sub(pointerdown);
+    obstacle.geometries[0]!.pose.orientationVector.th =
+      pointerdownTheta + pointermove.y;
+
+    // Scale
+  } else if (event.originalEvent.altKey) {
+    pointermove.set(event.point.x, event.point.y);
+    pointermove.sub(pointerdown);
+
+    const { y } = pointermove;
+
+    switch (obstacle.geometries[0]!.type) {
+      case 'sphere': {
+        obstacle.geometries[0]!.radius = Math.max(0, pointerdownRadius - y);
```

Comment:
This is mutating an object (kind deep in its graph) that is passed in as a prop. This seems like a relatively easy place for a bug to sneak into in the future.

As we discussed in the UI guild a couple weeks ago, what could we do here to avoid mutating objects that are passed around between components?

An immutable approach here feels a bit heavy-handed, given the depth of changes. Given that we want to notify the parent when this changes, would changing this to some sort of reactive store be better? Mutating inside a `set` or `update` isn't an issue, it could give us a helpful place to put some of this math, too, which would be easier to test, and I think it makes it more explicit that "this child will be changing the value inside this box"

Reply:
>  would changing this to some sort of reactive store be better

Yeah, I had considered this, I think your arguments here have finally convinced me, will update this!

Reply:
Hmmm switching to a store is causing more complexity / problems than it seems to be solving. Since prop editing is occuring only in this component I'm going to merge as is and schedule a session with you to revisit since these comments keep popping up in PRs and I'd like to find a good pattern that we both like :)

Reply:
Could the change event simply be the change data and then the parent can do the mutation?

Reply:
That makes sense, thanks for the breakdown! I update the component to read and write from the store value, obstacles are no longer passed from parent to child and then mutated here.